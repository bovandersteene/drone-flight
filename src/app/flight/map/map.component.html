<agm-map
  [latitude]="latitude"
  [longitude]="longitude"
  [zoom]="17"
  (mapClick)="addNode($event)"
  (centerChange)="centerChanged($event)"
>
  <agm-marker
    *ngFor="let path of paths; let i = index"
    [latitude]="path.lat"
    [longitude]="path.lng"
    [markerDraggable]="true"
    (markerClick)="deleteNode(i)"
    (dragEnd)="moveNode(i, $event)"
  ></agm-marker>
  <agm-polyline>
    <agm-polyline-point
      *ngFor="let path of paths"
      color="red"
      [latitude]="path.lat"
      [longitude]="path.lng"
    ></agm-polyline-point>
  </agm-polyline>
</agm-map>
