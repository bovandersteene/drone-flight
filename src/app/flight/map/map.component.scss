agm-map {
  height: calc(100% - 20px);
}
button {
  position: fixed;
  bottom: 20px;
  left: 20px;
  z-index: 3000;
}
mat-form-field {
  width: 100%;
}
