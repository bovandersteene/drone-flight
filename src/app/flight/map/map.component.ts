import {
  ChangeDetectionStrategy,
  Component,
  EventEmitter,
  Input,
  Output
} from "@angular/core";
import { LatLngLiteral } from "@agm/core";

@Component({
  selector: "app-map",
  templateUrl: "./map.component.html",
  styleUrls: ["./map.component.scss"],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class MapComponent {
  @Input()
  latitude: number;
  @Input()
  longitude: number;
  @Output()
  changePath = new EventEmitter<Array<LatLngLiteral>>();
  @Output()
  changeLocation = new EventEmitter<LatLngLiteral>();

  private _paths: Array<LatLngLiteral> = [];

  moveNode(index, coords: { coords: LatLngLiteral }) {
    this._paths[index] = coords.coords;
    this.changePath.emit(this.paths);
  }

  deleteNode(index) {
    this.paths.splice(index, 1);
    this.changePath.emit(this.paths);
  }

  addNode(coords: { coords: LatLngLiteral }) {
    this._paths = [...this.paths, coords.coords];
    this.changePath.emit(this.paths);
  }

  centerChanged(coords: LatLngLiteral) {
    this.changeLocation.emit(coords);
  }

  @Input()
  set paths(paths: Array<LatLngLiteral>) {
    this._paths = [...paths];
  }

  get paths() {
    return this._paths;
  }
}
