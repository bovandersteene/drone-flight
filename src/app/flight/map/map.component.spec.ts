import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { MapComponent } from "./map.component";
import { NO_ERRORS_SCHEMA } from "@angular/core";
import { describe } from "selenium-webdriver/testing";
import { LatLngLiteral } from "@agm/core";

const location = { lat: 46.5137907, lng: 6.6168853 } as LatLngLiteral;
const path = [
  { lat: 46.51952679624158, lng: 6.611394037382979 },
  { lat: 46.51926061859036, lng: 6.6125447888694 },
  { lat: 46.5184769874756, lng: 6.613104113268037 }
];

describe("MapComponent", () => {
  let component: MapComponent;
  let fixture: ComponentFixture<MapComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [MapComponent],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(MapComponent);
    component = fixture.componentInstance;
    component.latitude = location.lat;
    component.longitude = location.lng;
    component.paths = path;
    fixture.detectChanges();
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });
  it("should match the snapshot", () => {
    expect(fixture.debugElement.nativeElement).toMatchSnapshot();
  });
  describe("moveNode", () => {
    it("should call the change the node position", () => {
      jest.spyOn(component.changePath, "emit");
      component.moveNode(1, { coords: location });
      expect(component.changePath.emit).toHaveBeenCalledWith([
        path[0],
        location,
        path[2]
      ]);
    });
  });
  describe("deleteNode", () => {
    it("should call the delete the node", () => {
      jest.spyOn(component.changePath, "emit");
      component.deleteNode(1);
      expect(component.changePath.emit).toHaveBeenCalledWith([
        path[0],
        path[2]
      ]);
    });
  });
  describe("addNode", () => {
    it("should call the add the path to the pathlist", () => {
      jest.spyOn(component.changePath, "emit");
      component.addNode({ coords: location });
      expect(component.changePath.emit).toHaveBeenCalledWith([
        path[0],
        path[1],
        path[2],
        location
      ]);
    });
  });
  describe("centerChanged", () => {
    it("should change the location", () => {
      jest.spyOn(component.changeLocation, "emit");
      component.centerChanged(location);
      expect(component.changeLocation.emit).toHaveBeenCalledWith(location);
    });
  });
});
