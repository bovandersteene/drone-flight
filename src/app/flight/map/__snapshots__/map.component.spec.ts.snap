// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapComponent should match the snapshot 1`] = `
<div
  id="root1"
  ng-version="7.1.0"
>
  <agm-map>
    <!--bindings={
  "ng-reflect-ng-for-of": "[object Object],[object Object"
}-->
    <agm-marker />
    <agm-marker />
    <agm-marker />
    <agm-polyline>
      <!--bindings={
  "ng-reflect-ng-for-of": "[object Object],[object Object"
}-->
      <agm-polyline-point
        color="red"
      />
      <agm-polyline-point
        color="red"
      />
      <agm-polyline-point
        color="red"
      />
    </agm-polyline>
  </agm-map>
</div>
`;
